File: bindings/swift/Sources/Proven/SafeCalculator.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe expression evaluation delegated to libproven FFI.
///
/// Arithmetic expression parsing and evaluation via the formally
/// verified Idris 2 core.

import CProven

public enum SafeCalculator {
    /// Evaluate an arithmetic expression string.
    ///
    /// Supports: +, -, *, /, parentheses, negative numbers, decimals.
    public static func eval(_ expression: String) -> Result<Double, ProvenError> {
        withStringBytes(expression) { ptr, len in
            let result = proven_calculator_eval(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }
}

File: bindings/deno/src/safe_calculator.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeCalculator - Mathematical expression evaluation.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe expression calculator backed by formally verified Idris 2 code.
 */
export class SafeCalculator {
  /**
   * Evaluate a mathematical expression string.
   *
   * @param expression - The expression to evaluate (e.g. "2 + 3 * 4").
   * @returns Result containing the evaluated number or an error string.
   */
  static eval(expression: string): Result<number> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(expression);
    const result = symbols.proven_calculator_eval(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }
}
